src/auto-evo/simulation/MichePopulation.cs

```diff
@@ -141,157 +163,112 @@ private static List<Species> CopyInitialPopulationsToResults(SimulationConfigura
+                species.Where(s => parameters.Results.GetPopulationInPatch(s, entry.Value) > 0), random, cache);
         }
     }
 
     /// <summary>
     ///   The heart of the simulation that handles the processed parameters and calculates future populations.
     /// </summary>
     private static void SimulatePatchStep(SimulationConfiguration simulationConfiguration, Patch patch,
-        IEnumerable<Species> genericSpecies, Random random, SimulationCache cache,
-        IAutoEvoConfiguration autoEvoConfiguration, WorldGenerationSettings worldSettings)
+        IEnumerable<Species> genericSpecies, Random random, SimulationCache cache)
     {
         _ = random;
 
         var populations = simulationConfiguration.Results;
         bool trackEnergy = simulationConfiguration.CollectEnergyInformation;
-        bool dayNightCycle = worldSettings.DayNightCycleEnabled;
 
-        // This algorithm version is for microbe species
-        // TODO: add simulation for multicellular
-        var species = genericSpecies.Select(s => s as MicrobeSpecies).Where(s => s != null).Select(s => s!)
-            .ToList();
+        populations.MicheByPatch.TryGetValue(patch, out var miche);
+
+        foreach (var s in simulationConfiguration.ExtraSpecies)
+            miche!.InsertSpecies(s, cache);
+
+        var species = genericSpecies.Where(x => !simulationConfiguration.ExcludedSpecies.Contains(x))
+            .Union(simulationConfiguration.ExtraSpecies).Distinct().ToList();
 
         // Skip if there aren't any species in this patch
         if (species.Count < 1)
             return;
 
-        var energyBySpecies = new Dictionary<MicrobeSpecies, float>();
-        foreach (var currentSpecies in species)
-        {
-            energyBySpecies[currentSpecies] = 0.0f;
-        }
-
-        bool strictCompetition = autoEvoConfiguration.StrictNicheCompetition;
-
-        var niches = new List<FoodSource>
-        {
-            new EnvironmentalFoodSource(patch, Sunlight, Constants.AUTO_EVO_SUNLIGHT_ENERGY_AMOUNT),
-            new EnvironmentalFoodSource(patch, Temperature, Constants.AUTO_EVO_THERMOSYNTHESIS_ENERGY_AMOUNT),
-            new CompoundFoodSource(patch, Glucose, dayNightCycle),
-            new CompoundFoodSource(patch, HydrogenSulfide, dayNightCycle),
-            new CompoundFoodSource(patch, Iron, dayNightCycle),
-            new ChunkFoodSource(patch, "marineSnow"),
-            new ChunkFoodSource(patch, "ironSmallChunk"),
-            new ChunkFoodSource(patch, "ironBigChunk"),
-        };
-
-        foreach (var currentSpecies in species)
-        {
-            niches.Add(new HeterotrophicFoodSource(patch, currentSpecies, cache));
-        }
+        var leafNodes = miche!.GetLeafNodes().Where(x => x.Occupant != null).ToList();
+        var energyDictionary = species.ToDictionary(x => x, _ => 0.0);
 
-        foreach (var niche in niches)
+        foreach (var node in leafNodes)
```

Comment:
I see a bunch of enumerator and other memory allocations here, I'll need to get back to this later.